imap_processing/idex/IDEXPacketParser.py

```diff
@@ -0,0 +1,120 @@
+import bitstring
+import numpy as np
+from space_packet_parser import parser, xtcedef
+from imap_processing import packet_definition_directory
+
+
+class IDEXPacketParser():
+    def __init__(self, idex_packet_file: str):
+
+        idex_xtce = f"{packet_definition_directory}/idex_packet_definition.xml"
+        idex_definition = xtcedef.XtcePacketDefinition(xtce_document=idex_xtce)
+        idex_parser = parser.PacketParser(idex_definition)
+
+        idex_binary_data = bitstring.ConstBitStream(filename=idex_packet_file)
+        idex_packet_generator = idex_parser.generator(idex_binary_data)
+
+        self.epochs = []
+        self.data = {}
+        
+        evtnum = 0
+        for pkt in idex_packet_generator:
+            if 'IDX__SCI0TYPE' in pkt.data:
+                scitype = pkt.data['IDX__SCI0TYPE'].raw_value
+                if scitype == 1:
+                    evtnum += 1
+                    self.epochs.append(pkt.data['SHCOARSE'].derived_value + 20*(10**(-6))*pkt.data['SHFINE'].derived_value)  # Use this as the CDF epoch
+                    self._log_packet_info(evtnum, pkt) # These are for our logs, don't use
+
+                if scitype in [2, 4, 8, 16, 32, 64]:
+                    if scitype not in self.data:
+                        self.data.update({scitype : {}})
+                    if evtnum not in self.data[scitype]:
+                        self.data[scitype][evtnum] = pkt.data['IDX__SCI0RAW'].raw_value
+                    else:
+                        self.data[scitype][evtnum] += pkt.data['IDX__SCI0RAW'].raw_value
+
+        # Parse the waveforms according to the scitype present (high gain and low gain channels encode waveform data differently).
+        self.scitype_to_names = {2: "TOF_High", 4: "TOF_Low", 8: "TOF_Mid", 16: "Target_Low",
+                                32: "Target_High", 64: "Ion_Grid"}
+        datastore = {}
+        self.time_low_sr = []
+        self.time_high_sr = []
+        for scitype in self.data:
+            datastore[self.scitype_to_names[scitype]] = []
+            for evt in self.data[scitype]:
+                datastore[self.scitype_to_names[scitype]].append(self._parse_waveform_data(self.data[scitype][evt], scitype))
+                if self.scitype_to_names[scitype] == 'Target_Low':
+                    self.time_low_sr.append(np.linspace(0, len(datastore['Target_Low'][0]), len(datastore['Target_Low'][0])))
+                if self.scitype_to_names[scitype] == 'TOF_Low':
+                    self.time_high_sr.append(np.linspace(0, len(datastore['TOF_Low'][0]), len(datastore['TOF_Low'][0])))
+
+        self.data = datastore
+        self.numevents = evtnum
+    
+    def _log_packet_info(self, evtnum, pkt):
```

Comment:
I'm being nitpicky here, but could we use `event_num` and `packet` here instead of `evtnum` and `pkt`? I find that code is less readable when variables are unnecessarily abbreviated.

Reply:
yeah for sure